# MetaSBT

A scalable framework for automatically indexing microbial genomes and accurately characterizing metagenome-assembled genomes with Sequence Bloom Trees

Please visit the official [wiki](https://github.com/cumbof/MetaSBT/wiki) to get started.

> :warning: _This is still under development and some features may produce errors or unexpected results. You may also want to look at the [Contributing](https://github.com/cumbof/MetaSBT#contributing) and [Support](https://github.com/cumbof/MetaSBT#support) sections in case you will encounter a bug_

## Credits

Please credit our work in your manuscript by citing:

> _Manuscript in preparation_

## Contributing

Long-term discussion and bug reports are maintained via GitHub [Discussions](https://github.com/cumbof/MetaSBT/discussions) and [Issues](https://github.com/cumbof/MetaSBT/issues), while code review is managed via GitHub [Pull Requests](https://github.com/cumbof/MetaSBT/pulls).

Please, (i) be sure that there are no existing issues/PR concerning the same bug or improvement before opening a new issue/PR; (ii) write a clear and concise description of what the bug/PR is about; (iii) specifying the list of steps to reproduce the behavior in addition to versions and other technical details is highly recommended.

## Support

If you need support, please open an [Issue](https://github.com/cumbof/MetaSBT/issues) or a new [Discussion](https://github.com/cumbof/MetaSBT/discussions). We will be happy to answer your questions and help you troubleshoot any kind of issue concerning our framework.

Copyright © 2022 [Fabio Cumbo](https://github.com/cumbof), [Daniel Blankenberg](https://github.com/blankenberg). See [LICENSE](https://github.com/cumbof/MetaSBT/blob/main/LICENSE) for additional details.

[^1]: Grüning, Björn, et al. "[Bioconda: sustainable and comprehensive software distribution for the life sciences.](https://doi.org/10.1038/s41592-018-0046-7)" Nature methods 15.7 (2018): 475-476.
[^2]: Parks, Donovan H., et al. "[CheckM: assessing the quality of microbial genomes recovered from isolates, single cells, and metagenomes.](https://doi.org/10.1101/gr.186072.114)" Genome research 25.7 (2015): 1043-1055.
[^3]: Harris, Robert S., and Paul Medvedev. "[Improved representation of sequence bloom trees.](https://doi.org/10.1093/bioinformatics/btz662)" Bioinformatics 36.3 (2020): 721-727.
[^4]: Pornputtapong, Natapol, et al. "[KITSUNE: A Tool for Identifying Empirically Optimal K-mer Length for Alignment-Free Phylogenomic Analysis](https://doi.org/10.3389/fbioe.2020.556413)" Frontiers in Bioengineering and Biotechnology 8 (2020): 556413.
[^5]: Mohamadi, Hamid, Hamza Khan, and Inanc Birol. "[ntCard: a streaming algorithm for cardinality estimation in genomics data.](https://doi.org/10.1093/bioinformatics/btw832)" Bioinformatics 33.9 (2017): 1324-1330.
